winch/codegen/src/codegen/mod.rs

```diff
@@ -1417,6 +1424,146 @@ where
         }
         Ok(())
     }
+
+    #[cfg(not(feature = "threads"))]
+    pub fn emit_atomic_wait(&mut self, _arg: &MemArg, _kind: AtomicWaitKind) -> Result<()> {
+        Err(CodeGenError::unimplemented_wasm_instruction().into())
+    }
+
+    /// Emit the sequence of instruction for a `memory.atomic.wait*`.
+    #[cfg(feature = "threads")]
+    pub fn emit_atomic_wait(&mut self, arg: &MemArg, kind: AtomicWaitKind) -> Result<()> {
+        use crate::masm::ExtendKind;
+
+        // The `memory_atomic_wait*` builtins expect the following arguments:
+        // - `memory`, as u32
+        // - `address`, as u64
+        // - `expected`, as either u64 or u32
+        // - `timeout`, as u64
+        // At this point our stack only contains the `timeout`, the `expected` and the address, so
+        // we need to:
+        // - insert the memory as the first argument
+        // - compute the actual memory offset from the `MemArg`, if necessary.
+        // Note that the builtin function performs the alignment and bounds checks for us, so we
+        // don't need to emit that.
+
+        let timeout = self.context.pop_to_reg(self.masm, None)?;
+        let expected = self.context.pop_to_reg(self.masm, None)?;
+        let addr = self.context.pop_to_reg(self.masm, None)?;
+
+        // Put the target memory index in a register, and push it as the first argument.
+        let mem = self.context.any_gpr(self.masm)?;
+        self.masm.mov(
+            writable!(mem),
+            RegImm::i32(arg.memory as i32),
+            OperandSize::S32,
+        )?;
+        self.context
+            .stack
+            .push(TypedReg::new(WasmValType::I32, mem).into());
+
+        // compute the offset if necessary.
+        self.masm.extend(
+            writable!(addr.reg),
+            addr.reg,
+            ExtendKind::Unsigned(Extend::I64Extend32),
+        )?;
```

Comment:
I think this extend shouldn't be emitted unconditionally: in the case of `memory64`, the address is expected to be `64` bits, so there shouldn't be a need to emit this. 

Also, 32-bit loads are zero extend by default, so I think we can omit this extend entirely?

Reply:
Let me check that again, if I recall correctly I added it because it didn't work like I expected with negative offsets

Reply:
nevermind, I must have imagined that...